src/routes/_surfaces/workspace/explore/time-range-constructor/TimeGrainSelector.svelte

```diff
@@ -0,0 +1,105 @@
+  import MenuItem from "$lib/components/menu/MenuItem.svelte";
+  import FloatingElement from "$lib/components/tooltip/FloatingElement.svelte";
+  import { getMetricsExplorerById } from "$lib/redux-store/explore/explore-readables";
+  import { onClickOutside } from "$lib/util/on-click-outside";
+  import { tick } from "svelte";
+  import {
+    getDefaultTimeGrain,
+    getSelectableTimeGrains,
+    prettyTimeGrain,
+  } from "./timeRangeUtils";
+
+  export let metricsDefId: string;
+  export let selectedTimeRangeName: TimeRangeName;
+  export let selectedTimeGrain: TimeGrain;
+  export let onSelectTimeGrain: (timeGrain: TimeGrain) => void;
+
+  $: metricsExplorer = getMetricsExplorerById(metricsDefId);
+
+  let selectableTimeGrains: TimeGrain[];
+
+  // TODO: replace this with a call to the `/meta` endpoint, once available.
+  $: if (selectedTimeRangeName && $metricsExplorer?.allTimeRange) {
+    selectableTimeGrains = getSelectableTimeGrains(
+      selectedTimeRangeName,
+      $metricsExplorer.allTimeRange
+    );
+  }
+
+  // Set the default time grain when:
+  // - there's no time grain currently selected
+  // - the selected time grain is not in the list of selectable time grains (which can happen when the time range name is changed)
+  $: if (
+    !selectedTimeGrain ||
+    selectableTimeGrains.find(
+      (timeGrain) => timeGrain === selectedTimeGrain
+    ) === undefined
+  ) {
+    selectedTimeGrain = getDefaultTimeGrain(selectedTimeRangeName);
+    onSelectTimeGrain(selectedTimeGrain);
+  }
+
+  /// Start boilerplate for DIY Dropdown menu ///
+  let timeSelectorMenu;
+  let timeGrainMenuOpen = false;
+  let clickOutsideListener;
+  $: if (!timeGrainMenuOpen && clickOutsideListener) {
+    clickOutsideListener();
+    clickOutsideListener = undefined;
+  }
+
+  const buttonClickHandler = async () => {
+    timeGrainMenuOpen = !timeGrainMenuOpen;
+    if (!clickOutsideListener) {
+      await tick();
+      clickOutsideListener = onClickOutside(() => {
+        timeGrainMenuOpen = false;
+      }, timeSelectorMenu);
+    }
+  };
+
+  let target: HTMLElement;
+  /// End boilerplate for DIY Dropdown menu ///
+</script>
+
+<button
+  bind:this={target}
+  class="px-4 py-2 rounded flex flex-row gap-x-2 hover:bg-gray-200 transition-tranform duration-100"
+  on:click={buttonClickHandler}
+>
+  <div>
+    by <span class="font-bold">{prettyTimeGrain(selectedTimeGrain)}</span> increments
+  </div>
+  <span class="transition-transform" class:-rotate-180={timeGrainMenuOpen}>
+    <CaretDownIcon size="16px" />
+  </span>
+</button>
+
+{#if timeGrainMenuOpen}
+  <div bind:this={timeSelectorMenu}>
+    <FloatingElement
+      relationship="direct"
+      location="bottom"
+      alignment="start"
+      {target}
+      distance={8}
+    >
+      <Menu on:escape={() => (timeGrainMenuOpen = false)}>
+        {#each selectableTimeGrains as timeGrain}
+          <MenuItem on:select={() => onSelectTimeGrain(timeGrain)}>
```

Comment:
This is where you'd actually call `dispatch('select-time-grain')`. In the component JS,

```
import {createEventDispatcher} from "svelte";
const dispatch = createEventDispatcher()
```

Then use `dispatch(eventName, args)` where needed.